### PR TITLE
Self-serve categories

### PR DESCRIPTION
**Pull request by Lukewh**
Fixes https://github.com/canonical-websites/snapcraft.io/issues/1458

## Done
- Added select boxes and interaction as described in https://docs.google.com/document/d/1Teovi34AVAkHtFdO9agsILqS96QVmaDqZ9DMKyFs48U/edit
- Updated the backend to transform the snap categories to avoid reserved jinja2 key

## QA
- Pull the branch
- `./run`
- Visit http://0.0.0.0:8004/snaps
- Select a snap
- You should see a new `Category:` section
- Play around with the categories, adding categories and secondary categories
- You should not be able to select the same category for both the primary and secondary categories
- When you're happy click "Save"
- Visit http://0.0.0.0:8004/search?category=<category_name> for the category(s) you selected
- If you go to the last page you should see your snap (may require patience and refreshes)
- **WHEN DONE, REMOVE THE CATEGORIES FROM TEST SNAPS AS THEY WILL DISPLAY ON PRODUCTION**

### CODE REVIEW

**Comment by therealjuan** [comment]
Works like a charm! 👍

**Comment by tbille** [comment]
@Lukewh 

The problem is that the API doesn't have a notion of Primary or Secondary but only alphabetical...

To reproduce the problem:

- https://snapcraft-io-canonical-websites-pr-1479.run.demo.haus/toto/listing

```
Primary: Social networking
Secondary: Productivity
```

- Save and reload
- The categories are swapped

**Comment by tbille** [comment]
QA fun:

```
Primary: Productivity
Second: Social networking
```

Remove second category, and replace first category with the deleted category.

We get an error: `Category names not found: {''} `

File: static/js/publisher/form.js [hunk 1]
```diff
@@ -113,14 +114,15 @@ function initForm(config, initialState, errors) {
   let state = JSON.parse(JSON.stringify(initialState));
 
   const stateInput = document.createElement("input");
-  stateInput.type = "hidden";
```
Comment (bartaz) [inline static/js/publisher/form.js]: Was it removed for debugging?

screen shot 2018-12-20 at 17 20 33

Comment (Lukewh) [reply static/js/publisher/form.js]: ahh good catch, didn't mean to remove it!

File: static/js/publisher/form.test.js [hunk 2]
```diff
@@ -34,3 +35,236 @@ describe("initSnapIconEdit", () => {
...
+
+    beforeEach(() => {
+      setupForm(config, initialState);
+    });
+
+    afterEach(() => {
+      form.parentNode.removeChild(form);
+    });
+
+    test("creates state input", () => {
```
[the remark this answers is not in the document]
Comment (Lukewh) [reply static/js/publisher/form.test.js]: I try and avoid "should" to be a more definitive statement. like "form when initialized creates a state input" :man_shrugging:

File: static/js/publisher/form.test.js [hunk 3]
```diff
@@ -34,3 +35,236 @@ describe("initSnapIconEdit", () => {
...
+    form.appendChild(websiteInput);
+    form.appendChild(contactInput);
+
+    document.body.appendChild(form);
+
+    categories.categories = jest.fn();
+    market.initForm(config, initialState, undefined);
+  }
+
+  describe("simple", () => {
```
[the remark this answers is not in the document]
Comment (Lukewh) [reply static/js/publisher/form.test.js]: badly named?

I was just hacking in some tests so codecov would chill out :upside_down_face:

File: static/js/publisher/form.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lukewh) [reply static/js/publisher/form.test.js]: https://github.com/canonical-websites/snapcraft.io/issues/1482

[the remark this answers is not in the document]
Comment (bartaz) [reply static/js/publisher/form.test.js]: It's not like we have any guidelines for that, so I don't mind. Would be good to have some consistency around that. Anyway, better to have some tests inconsistently named then none, so thanks for taking a time to add those! 👍

File: static/js/publisher/form.test.js [hunk 4]
```diff
@@ -34,3 +35,236 @@ describe("initSnapIconEdit", () => {
...
+    form.appendChild(primaryCategoryInput);
+    form.appendChild(secondaryCategoryInput);
+    form.appendChild(summaryInput);
+    form.appendChild(descriptionInput);
+    form.appendChild(websiteInput);
+    form.appendChild(contactInput);
+
+    document.body.appendChild(form);
+
+    categories.categories = jest.fn();
```
Comment (bartaz) [inline static/js/publisher/form.test.js]: I'm not sure why this works at all.
We are importing a module, changing its contents to a mock and expecting that it will be mocked when imported by another module (form) that we imported even earlier.

The fact that it works means that `categories` is now mocked for anything that would try to import it. So in theory it could affect other tests that import `categories`. But for some reason it doesn't. Maybe it's because of the order tests are run in, maybe because of a way `jest` runs tests in paralel and they don't share `categories` import between tests but it is shared within the test here? - I'm not sure.

Long story short - I think it's working by accident and it's possible it may have unexpected side effects (may modify `categories` for other tests).
Mocking imports is unfortunately hard - we used https://github.com/thlorenz/proxyquire for that in BSI.

I think I'd rather remove this mock and test before it causes any trouble :)

File: templates/publisher/listing.html [hunk 5]
```diff
@@ -140,6 +140,93 @@
           </div>
         </div>
 
+        <div class="row">
+          <div class="col-2">
+            <label>Category:</label>
+          </div>
+          <div class="col-4">
+            <input type="hidden" name="categories"
```
Comment (bartaz) [inline templates/publisher/listing.html]: Can't it be handled in the `state` directly? Do we need this additional hidden input for categories data?
We are pushing exactly the same data into state via JS couple of lines below in the template.

Comment (Lukewh) [reply templates/publisher/listing.html]: :+1: good point https://github.com/canonical-websites/snapcraft.io/issues/1483

File: static/js/publisher/form.test.js (hunk 4 above)
Comment (Lukewh) [reply static/js/publisher/form.test.js]: https://github.com/canonical-websites/snapcraft.io/issues/1484

Comment (Lukewh) [reply static/js/publisher/form.test.js]: Thanks
